core/merkle-tree.md

```diff
@@ -0,0 +1,133 @@
+# Binary Merkle Tree
+
+## Usage
+
+The binary hash tree facilitates succinct equality and item membership checks
+of constant-size list of binary blobs without disclosing the contents of the array itself.
+
+The input is "accumulated" to a single 32 byte root hash by recursively hashing pairs of two nodes into one.
+
+## Structure
+
+- Each tree node is identified by a SHA-256 hash.
+- Two types of nodes exist: *Leaf* and *intermediate* nodes.
+  - The pre-image of a *leaf node* is the byte `0x00` followed by an arbitrary amount of data.
+  - The pre-image of an *intermediate node* is the byte `0x01` followed by two 32 byte hashes, each referring to a node.
+- Each node has one or zero parent intermediate nodes.
+  - Referring to the same node twice in the same intermediate node is permitted.
+- The graph of nodes represents a binary tree, thus is acyclic and has one *root node*.
+
+## Algorithms
+
+### Leaf Order
+
+The *leaf order* of a tree is defined by depth-first search traversal starting at the root,
+counting only leaf nodes.
+
+**Example**
+
+The leaf order of the tree in _figure 1_ is `[Lβ, Lα, Lδ]`.
+
+<a id="figure_1"></a>
+
+```
+Figure 1: Non-canonical tree with three leaf nodes and two intermediate nodes
+
+     Iε
+    /  \
+   Iγ  Lδ
+  /  \
+ Lβ  Lα
+```
+
+### Level Order
+
+The *level order* of a tree is defined by depth-first search starting at the root,
+counting only nodes with a given level.
+
+**Example**
+
+The tree in [_figure 1_](#figure_1) has the following level orders:
+ - `0: [Iε]`
+ - `1: [Iγ, Lδ]`
+ - `2: [Lβ, Lα]`
+
+### Canonical Construction
```

Comment:
This term is new and doesn't appear in the Labs repo, should review whether we want to call it as such